File: 6_pilas_colas/parentesisValidos.cpp
//Programando Parentesis Válido con C++

bool isValid(string s){
    stack<char> pila;
    for (char& caracter : s){
        if (caracter == ')'){
            if(pila.empty()){
                return false;
            }
            pila.pop();
        }
        else if (caracter == '('){
            pila.push('(');
        }
    }
    return pila.empty();
}

// ())(  stack -> return false
//((abc)*d) stact -> true  //complejida temporal O(n) - complejidad espacial O(n)

File: 6_pilas_colas/parentesisValidos.java
import java.util.Stack;

public class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == ')') {
                if (stack.empty()) {
                    return false;
                }
                stack.pop();
            } else if (c == '(') {
                stack.push('(');
            }
        }
        return stack.empty();
    }
}